### PR TITLE
fix(meta): do not check upload count for heartbeat

### PR DESCRIPTION
**Pull request by Little-Wallace**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

### <samp>🤖 Generated by Copilot at b9cd773</samp>

This pull request enhances the compaction group manager and the compactor manager in the Hummock storage engine. It improves the efficiency and correctness of table removal and compaction progress tracking. It also adds some helper methods and functions to the `HummockVersionExt` trait and the `hummock_version_ext.rs` file.

### <samp>🤖 Generated by Copilot at b9cd773</samp>

*  Simplify the logic of updating the task expiry time based on the progress report ([link](https://github.com/risingwavelabs/risingwave/pull/9719/files?diff=unified&w=0#diff-dfe10b5f95ee1a2e8cd6881d78ad0c5b80522b986da8efee93858c644ccdc8d2L401-R402))
*  Modify the type of the `modified_groups` map to store the table IDs that are removed from each group ([link](https://github.com/risingwavelabs/risingwave/pull/9719/files?diff=unified&w=0#diff-662a20379f373bb7b7fe43536238deac2c80e955dd463c592cd2cf4bf6bf0b05L281-R281))
*  Handle the case of partial removal of tables from a group and GC the associated SSTs ([link](https://github.com/risingwavelabs/risingwave/pull/9719/files?diff=unified&w=0#diff-662a20379f373bb7b7fe43536238deac2c80e955dd463c592cd2cf4bf6bf0b05L303-R345), [link](https://github.com/risingwavelabs/risingwave/pull/9719/files?diff=unified&w=0#diff-662a20379f373bb7b7fe43536238deac2c80e955dd463c592cd2cf4bf6bf0b05R354-R358), [link](https://github.com/risingwavelabs/risingwave/pull/9719/files?diff=unified&w=0#diff-1e8919a212b518d9a3fccef41f58da25799a5d99ae3dfeda752723651127a15fR141-R150), [link](https://github.com/risingwavelabs/risingwave/pull/9719/files?diff=unified&w=0#diff-1e8919a212b518d9a3fccef41f58da25799a5d99ae3dfeda752723651127a15fR427-R470), [link](https://github.com/risingwavelabs/risingwave/pull/9719/files?diff=unified&w=0#diff-1e8919a212b518d9a3fccef41f58da25799a5d99ae3dfeda752723651127a15fR702-R739))
  *  Use the `HummockVersionExt` trait to collect and clear the dropped SST infos from the `HummockVersion` struct ([link](https://github.com/risingwavelabs/risingwave/pull/9719/files?diff=unified&w=0#diff-1e8919a212b518d9a3fccef41f58da25799a5d99ae3dfeda752723651127a15fR141-R150), [link](https://github.com/risingwavelabs/risingwave/pull/9719/files?diff=unified&w=0#diff-1e8919a212b518d9a3fccef41f58da25799a5d99ae3dfeda752723651127a15fR427-R470), [link](https://github.com/risingwavelabs/risingwave/pull/9719/files?diff=unified&w=0#diff-1e8919a212b518d9a3fccef41f58da25799a5d99ae3dfeda752723651127a15fR702-R739))
  *  Update the branched SSTs map and the GC object IDs list with the dropped SSTs ([link](https://github.com/risingwavelabs/risingwave/pull/9719/files?diff=unified&w=0#diff-662a20379f373bb7b7fe43536238deac2c80e955dd463c592cd2cf4bf6bf0b05L303-R345))
  *  Update the intra-level delta to reflect the removal of the SSTs from the group ([link](https://github.com/risingwavelabs/risingwave/pull/9719/files?diff=unified&w=0#diff-662a20379f373bb7b7fe43536238deac2c80e955dd463c592cd2cf4bf6bf0b05L303-R345))
  *  Clear the dropped SST infos from the current version after applying the new version delta ([link](https://github.com/risingwavelabs/risingwave/pull/9719/files?diff=unified&w=0#diff-662a20379f373bb7b7fe43536238deac2c80e955dd463c592cd2cf4bf6bf0b05R354-R358))

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/meta/src/hummock/manager/compaction_group_manager.rs [hunk 1]
```diff
@@ -300,45 +299,50 @@ impl<S: MetaStore> HummockManager<S> {
...
-            .into_iter()
-            .filter_map(|(group_id, member_count)| {
-                if member_count == 0 && group_id > StaticCompactionGroupId::End as CompactionGroupId
+        let mut dirty_groups = vec![];
+        for (group_id, mut removed_table_ids) in modified_groups {
+            if let Some(group) = versioning.current_version.levels.get(&group_id) {
+                removed_table_ids.sort();
+                let mut dropped_sstables = vec![];
+                if group_id > StaticCompactionGroupId::End as CompactionGroupId
+                    && group.member_table_ids.eq(&removed_table_ids)
```
Comment (zwang28) [inline src/meta/src/hummock/manager/compaction_group_manager.rs:314]: member_table_ids is not guaranteed to be sorted, e.g. [here](https://github.com/risingwavelabs/risingwave/blob/a2ea8d4c80cb6bd3d65f91f0c0775011958fff10/src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs#L456). Not sure if any other places.

**Review by zwang28 — APPROVED** [review]
LGTM for the rest.

File: src/meta/src/hummock/manager/compaction_group_manager.rs (hunk 1 above)
Comment (Little-Wallace) [reply src/meta/src/hummock/manager/compaction_group_manager.rs:314]: No. We sort it every time after we change it

**Comment by Li0k** [comment]
Do we need to clean up the ssts corresponding to drop in the meta? Is filtering on the compactor side sufficient?

The reason is: the removal of ssts from version is currently triggered by report_compaction_task. However, in this pr we have added a new path. Will it be difficult to maintain in the future? @hzxa21 @Little-Wallace @zwang28

**Comment by Li0k** [comment]
We have two options to reduce the io of compactor due to out-of-date data
1. filter expired data on the compactor side
2. as pr modification

For pr modifications, we can clear the sst directly from the version, so that we can free up space faster. However, this cleanup happens at the same time as the compaction, when report_compact_task may find that the input sst has been deleted, how do we handle this (determine is_compact_task_expired ?)

Discussion:
- If we clean up the sst directly in the version, which already does the work of space reclaim, can we remove the space reclaim compaction ??
- Is it possible to replace the space reclaim work after splitting

File: src/storage/hummock_test/src/compactor_tests.rs [hunk 2]
```diff
@@ -550,12 +550,6 @@ pub(crate) mod tests {
         )
         .await;
 
-        let compact_ctx_existing_table_id = prepare_compactor_and_filter(
```
Comment (Li0k) [inline src/storage/hummock_test/src/compactor_tests.rs:553]: why do we remove this code ?

**Review by Li0k — APPROVED** [review]
Rest LGTM

**Pull request merged**